pkg/ds/farm.go

```diff
@@ -0,0 +1,342 @@
+	"github.com/square/p2/pkg/kp/dsstore"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/scheduler"
+	"github.com/square/p2/pkg/types"
+	klabels "k8s.io/kubernetes/pkg/labels"
+)
+
+// Farm instatiates and deletes daemon sets as needed
+type Farm struct {
+	// constructor arguments
+	kpStore    kp.Store
+	dsStore    dsstore.Store
+	scheduler  scheduler.Scheduler
+	applicator labels.Applicator
+
+	children map[fields.ID]*childDS
+	childMu  sync.Mutex
+
+	logger  logging.Logger
+	alerter alerting.Alerter
+}
+
+type childDS struct {
+	ds        DaemonSet
+	quitCh    chan<- struct{}
+	updatedCh chan<- *ds_fields.DaemonSet
+	deletedCh chan<- *ds_fields.DaemonSet
+	errCh     <-chan error
+}
+
+func NewFarm(
+	kpStore kp.Store,
+	dsStore dsstore.Store,
+	applicator labels.Applicator,
+	logger logging.Logger,
+) *Farm {
+	return &Farm{
+		kpStore:    kpStore,
+		dsStore:    dsStore,
+		scheduler:  scheduler.NewApplicatorScheduler(applicator),
+		applicator: applicator,
+		logger:     logger,
+		children:   make(map[fields.ID]*childDS),
+	}
+}
+
+func (dsf *Farm) Start(quitCh <-chan struct{}) {
+	dsf.cleanupDaemonSetPods(quitCh)
+	dsf.mainLoop(quitCh)
+}
+
+const cleanupInterval = 60 * time.Second
+
+// This function removes all pods with a DSIDLabel where the daemon set id does
+// not exist in the store at every interval specified because it is possible
+// that the farm will unexpectedly crash or someone deletes or modifies a node
+func (dsf *Farm) cleanupDaemonSetPods(quitCh <-chan struct{}) {
+	go func() {
+		timer := time.NewTimer(time.Duration(0))
+
+		for {
+			select {
+			case <-quitCh:
+				return
+			case <-timer.C:
+			}
+			timer.Reset(cleanupInterval)
+
+			allDaemonSets, err := dsf.dsStore.List()
+			dsIDMap := make(map[string]ds_fields.DaemonSet)
+			for _, dsFields := range allDaemonSets {
+				dsIDMap[dsFields.ID.String()] = dsFields
+			}
+
+			allPods, err := dsf.applicator.GetMatches(klabels.Everything(), labels.POD)
+			if err != nil {
+				dsf.logger.NoFields().Fatal(err)
+				return
+			}
+
+			for _, podLabels := range allPods {
+				// Only check if it is a pod scheduled by a daemon set
+				if podLabels.Labels.Has(DSIDLabel) {
+					dsID := podLabels.Labels.Get(DSIDLabel)
+
+					// Check if the daemon set exists, if it doesn't unschedule the pod
+					if _, ok := dsIDMap[dsID]; !ok {
+						nodeName, podID, err := labels.NodeAndPodIDFromPodLabel(podLabels)
+						if err != nil {
+							dsf.logger.NoFields().Fatal(err)
+							return
+						}
+
+						// TODO: Since this mirrors the unschedule function in daemon_set.go,
+						// We should find a nice way to couple them together
+						dsf.logger.NoFields().Infof("Unscheduling from %s with dangling daemon set uuid %s", nodeName, dsID)
+
+						_, err = dsf.kpStore.DeletePod(kp.INTENT_TREE, nodeName, podID)
+						if err != nil {
+							dsf.logger.NoFields().Errorf("Unable to delete pod from intent tree: %v", err)
```

Comment:
Is the error handling strategy here something like: "We'll attempt this same DeletePod call on the next iteration of the loop if we can't delete it now for some reason?"

That's fine, but I'd like to see an ErrorReporter call happen here also.